opendevin/core/main.py

```diff
@@ -33,14 +37,15 @@ def read_task_from_stdin() -> str:
     return sys.stdin.read()
 
 
-async def run_agent_controller(
-    agent: Agent,
+async def run_controller(
+    config: AppConfig,
```

Comment:
Can we avoid passing in `AppConfig` here? In general, I think it's best to avoid passing in big config classes when it can be at all avoided as it obscures which arguments are actually necessary for this function and also obscures the function behavior (the behavior of this function is significantly more complex than before).

Here, it seems that we use:
- `config.default_agent`: we're already passing in `Agent`, not necessary
- `config.max_iterations`: we're already passing in `max_iterations`, not necessary
- `config.get_llm_config_from_agent(config.default_agent)`: only used when `Agent` is none, but if we don't have `None` agents it's not necessary

Reply:
The reason is that `run_controller` will be our main point of entry since it will initialize agent, event stream, and the runtime.

In evaluation, we typically need to customize *a lot* of stuff; if we don't pass the `AppConfig,` we may end up passing tens of different configs to `run_controller`, which could obscure things even more (`AppConfig` is at least structured).

For example, here's my work-in-progress for the SWE-Bench eval refactor; we need to change a bunch of `sandbox` behavior while customizing container images. I was thinking passing one `AppConfig` could be a relatively cleaner way.. WDYT?

https://github.com/OpenDevin/OpenDevin/blob/c3d6320f381c93e032598ea54bbeddfac4e83ac2/evaluation/swe_bench/run_infer.py#L117-L144

Reply:
Is this actually simpler than the alternative? It seems that you still need to pass in arguments to AppConfig, so why not just pass them in to run_controller?

Reply:
To explain a little bit more, in the previous code it had the following behavior:

1. use Agent

now the new code has this behavior

1. if Agent is not None, use Agent
2. if Agent is None, find the default LLM config for the config.default_agent, initialize config.default_agent with that LLM and use it

The "2" behavior is complex and unpredictable to people who are not familiar with the function, and could be a source of error in programming or experiments. I think it'd be nice to keep things simple and easily understandable if we can.

Reply:
I think the need for tweaking these different arguments will likely increase as we incorporate more benchmarks (e.g., each requires a slightly different setting). If we don't have a centralized structured place like `AppConfig` to store all these configurations, people may end-up keep adding arguments to `run_controller`.

And as you see, we need a lot of stuffs from `config` for this `run_controller` function, because it is the main entrypoint of OpenDevin:

![8a025df6666c6f5ad2a7f29c588fa141](https://github.com/user-attachments/assets/2a719fdf-5235-4f7e-b5e4-c2f52f7edfd7)

Reply:
And before the modifications made in this PR, `config = load_app_config()` is actually used in the file context to load a "global config" and having the input argument overwriting them.

In that original implementation, people actually need to change their `config.toml` to run eval (because `config = load_app_config()` load from `config.toml` and env var) -- And it could happen that we modify the config classes so the original `config.toml` that used to work for eval no longer works (This happens to me before: #3074).

https://github.com/OpenDevin/OpenDevin/pull/3184/files#diff-b764c5398bce883365e8821577ee6017565697e91e62dc06ef0c63b250f26acaL22-L23

Reply:
With the changes of requiring a `AppConfig` as input argument, i think it will actually be clearer to people what configuration is used for evaluation. Because values are hard-coded for different eval, it will make it easier to reproduce eval runs (e.g., if i accidentally put `use_auto_lint` in a different section like `[llm]` in `config.toml`, it may drops my score without me knowing it).

Reply:
@xingyaoww and I had a quick chat offline, and I realize that I was misunderstanding the previous state of the code. I'm OK with this change!